batcher/sweep_batcher.go

```diff
@@ -0,0 +1,286 @@
+
+	// musig2ServerKit includes all the required functionality to collect
+	// and verify signatures by the swap server in order to cooperatively
+	// sweep funds.
+	musig2ServerSign MuSig2SignSweep
+
+	// store includes all the database interactions that are needed by the
+	// batcher and the batches.
+	store BatcherStore
+}
+
+// NewBatcher creates a new Batcher instance.
+func NewBatcher(wallet lndclient.WalletKitClient,
+	chainNotifier lndclient.ChainNotifierClient,
+	signerClient lndclient.SignerClient, musig2ServerSigner MuSig2SignSweep,
+	store BatcherStore) *Batcher {
+
+	return &Batcher{
+		batches:          make(map[lntypes.Hash]*Batch),
+		SweepReqs:        make(chan *Sweep),
+		errChan:          make(chan error),
+		completeChan:     make(chan lntypes.Hash),
+		wallet:           wallet,
+		chainNotifier:    chainNotifier,
+		signerClient:     signerClient,
+		musig2ServerSign: musig2ServerSigner,
+		store:            store,
+	}
+}
+
+// Run starts the batcher and processes incoming sweep requests.
+func (br *Batcher) Run(ctx context.Context) error {
+	// First we fetch all the batches that are not in a confirmed state from
+	// the database. We will then start these batches and add them to our
+	// map of batches.
+	batches, err := br.FetchUnconfirmedBatches(ctx)
+	if err != nil {
+		return err
+	}
+
+	for _, batch := range batches {
+		_, err := br.spinUpBatchFromDB(ctx, *batch)
+		if err != nil {
+			return err
+		}
+	}
+
+	for {
+		select {
+		case sweep := <-br.SweepReqs:
+			err := br.handleSweep(ctx, sweep)
+			if err != nil {
+				return err
+			}
+
+		case id := <-br.completeChan:
+			delete(br.batches, id)
+
+		case err := <-br.errChan:
+			return err
+
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+// handleSweep handles a sweep request by either placing it in an existing
+// batch, or by spinning up a new batch.
+func (br *Batcher) handleSweep(ctx context.Context, sweep *Sweep) error {
+	// Check if the sweep is already in a batch. If that is the case, we
+	// provide the sweep to that batch and return.
+	for _, batch := range br.batches {
+		if batch.SweepExists(sweep.SwapHash) {
+			batch.RequestChan <- sweep
+			return nil
+		}
+	}
+
+	// If one of the batches accepts the sweep, we provide it to that batch.
+	for _, batch := range br.batches {
+		if batch.AcceptsSweep(sweep) {
+			batch.RequestChan <- sweep
+			return nil
+		}
+	}
+
+	// If no batch is capable of accepting the sweep, we spin up a fresh
+	// batch and hand over the sweep to it.
+	batch, err := br.spinUpBatch(ctx)
+	if err != nil {
+		return err
+	}
+
+	select {
+	case batch.RequestChan <- sweep:
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+// spinUpBatch spins up a new batch and returns it.
+func (br *Batcher) spinUpBatch(ctx context.Context) (*Batch, error) {
+	cfg := BatchConfig{
+		MaxTimeoutDistance: DefaultMaxTimeoutDistance,
+		BatchConfTarget:    defaultBatchConfTarget,
+	}
+
+	batch, err := NewBatch(
```

Comment:
nit: could use a functional options pattern or a config here to help readability.